File: Design-Patterns-Using-Java/ArchitecturalPatterns/PipeAndFilter.java

import java.util.ArrayList;
import java.util.List;

// Filter Base Class
abstract class Filter {
    abstract String process(String data);
}

// Filters
class CapitalizeFilter extends Filter {
    @Override
    String process(String data) {
        return data.toUpperCase();
    }
}

class ReplaceSpaceFilter extends Filter {
    @Override
    String process(String data) {
        return data.replace(" ", "_");
    }
}

class RemoveSpecialCharactersFilter extends Filter {
    @Override
    String process(String data) {
        String specialCharacters = ",@!";
        StringBuilder result = new StringBuilder();
        for (char c : data.toCharArray()) {
            if (specialCharacters.indexOf(c) == -1) {
                result.append(c);
            }
        }
        return result.toString();
    }
}

// Data Processing Pipeline
class DataProcessingPipeline {
    private List<Filter> filters;

    public DataProcessingPipeline() {
        this.filters = new ArrayList<>();
    }

    public void addFilter(Filter filter) {
        this.filters.add(filter);
    }

    public String processData(String data) {
        for (Filter filter : this.filters) {
            data = filter.process(data);
        }
        return data;
    }
}

// Main class
public class PipeAndFilter {
    public static void main(String[] args) {
        DataProcessingPipeline pipeline = new DataProcessingPipeline();
        pipeline.addFilter(new CapitalizeFilter());
        pipeline.addFilter(new ReplaceSpaceFilter());
        pipeline.addFilter(new RemoveSpecialCharactersFilter());

        String data = "Hello, World!";
        String result = pipeline.processData(data);
        System.out.println("Result: " + result);  // Output: "HELLO_WORLD"
    }
}

File: Design-Patterns-Using-Kotlin/ArchitecturalPatterns/PipeAndFilter.kt
import java.util.ArrayList

// Filter Base Class
abstract class Filter {
    abstract fun process(data: String): String
}

// Filters
class CapitalizeFilter : Filter() {
    override fun process(data: String): String {
        return data.uppercase()
    }
}

class ReplaceSpaceFilter : Filter() {
    override fun process(data: String): String {
        return data.replace(" ", "_")
    }
}

class RemoveSpecialCharactersFilter : Filter() {
    override fun process(data: String): String {
        val specialCharacters = ",@!"
        val result = StringBuilder()
        for (c in data.toCharArray()) {
            if (!specialCharacters.contains(c)) {
                result.append(c)
            }
        }
        return result.toString()
    }
}

// Data Processing Pipeline
class DataProcessingPipeline {
    private val filters: MutableList<Filter> = ArrayList()

    fun addFilter(filter: Filter) {
        filters.add(filter)
    }

    fun processData(data: String): String {
        var processedData = data
        for (filter in filters) {
            processedData = filter.process(processedData)
        }
        return processedData
    }
}

// Client code.
fun main() {
    val pipeline = DataProcessingPipeline()
    pipeline.addFilter(CapitalizeFilter())
    pipeline.addFilter(ReplaceSpaceFilter())
    pipeline.addFilter(RemoveSpecialCharactersFilter())

    val data = "Hello, World!"
    val result = pipeline.processData(data)
    println("Result: $result")  // Output: "HELLO_WORLD"
}
